include/DDpackage.hpp

```diff
@@ -0,0 +1,2047 @@
+
+                std::string line;
+                std::string complex_real_regex = R"(([+-]?(?:\d+(?:\.\d*)?|\.\d+)(?:[eE][+-]?\d+)?(?![ \d\.]*(?:[eE][+-])?\d*[iI]))?)";
+                std::string complex_imag_regex = R"(( ?[+-]? ?(?:(?:\d+(?:\.\d*)?|\.\d+)(?:[eE][+-]?\d+)?)?[iI])?)";
+                std::string edge_regex         = " \\(((-?\\d+) (" + complex_real_regex + complex_imag_regex + "))?\\)";
+                std::regex  complex_weight_regex(complex_real_regex + complex_imag_regex);
+                std::string line_construct = "(\\d+) (\\d+)";
+                for (auto i = 0U; i < N; ++i) {
+                    line_construct += "(?:" + edge_regex + ")";
+                }
+                line_construct += " *(?:#.*)?";
+                std::regex  line_regex(line_construct);
+                std::smatch m;
+
+                if (std::getline(is, line)) {
+                    if (!std::regex_match(line, m, complex_weight_regex)) {
+                        throw std::runtime_error("Regex did not match second line: " + line);
+                    }
+                    rootweight.from_string(m.str(1), m.str(2));
+                }
+
+                while (std::getline(is, line)) {
+                    if (line.empty() || line.size() == 1) continue;
+
+                    if (!std::regex_match(line, m, line_regex)) {
+                        throw std::runtime_error("Regex did not match line: " + line);
+                    }
+
+                    // match 1: node_idx
+                    // match 2: qubit_idx
+
+                    // repeats for every edge
+                    // match 3: edge content
+                    // match 4: edge_target_idx
+                    // match 5: real + imag (without i)
+                    // match 6: real
+                    // match 7: imag (without i)
+                    node_index = std::stoi(m.str(1));
+                    v          = static_cast<Qubit>(std::stoi(m.str(2)));
+
+                    for (auto edge_idx = 3U, i = 0U; i < N; i++, edge_idx += 5) {
+                        if (m.str(edge_idx).empty()) continue;
+
+                        edge_indices[i] = std::stoi(m.str(edge_idx + 1));
+                        edge_weights[i].from_string(m.str(edge_idx + 3), m.str(edge_idx + 4));
+                    }
+
+                    result = deserializeNode(node_index, v, edge_indices, edge_weights, nodes);
+                }
+            }
+
+            auto w = cn.getCached(rootweight.r, rootweight.i);
+            ComplexNumbers::mul(w, result.w, w);
+            result.w = cn.lookup(w);
+            cn.returnToCache(w);
+
+            return result;
+        }
+
+        template<class Node, class Edge = Edge<Node>>
+        Edge deserialize(const std::string& inputFilename, bool readBinary) {
+            auto ifs = std::ifstream(inputFilename);
+
+            if (!ifs.good()) {
+                std::cerr << "Cannot open serialized file: " << inputFilename << std::endl;
```

Comment:
I'd rather the app throws/crashes early instead of returning a default.

Reply:
Agreed. Now throws an `std::invalid_argument`.